internal/provider/provider.go

```diff
@@ -109,59 +71,101 @@ func (p *provider) login(ctx context.Context, client *api.Client, vaultKubernete
+	queryIndex := strings.Index(secretPath, "?")
+	var queryParams map[string][]string
+	if queryIndex != -1 {
+		var err error
+		queryParams, err = url.ParseQuery(secretPath[queryIndex+1:])
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse query parameters from secretPath %q for objectName %q: %w", secretPath, secret.ObjectName, err)
+		}
+		secretPath = secretPath[:queryIndex]
 	}
-
-	endpoint, err := generateSecretEndpoint(secretMountType, secretMountVersion, secretPrefix, secretSuffix)
-	if err != nil {
-		return "", err
+	method := "GET"
+	if secret.Method != "" {
+		method = secret.Method
 	}
 
-	p.logger.Debug("vault: Requesting valid secret mounted", "endpoint", endpoint)
-
-	req := client.NewRequest("GET", endpoint)
-	if secret.ObjectVersion != "" {
-		req.Params = url.Values{
-			"version": []string{secret.ObjectVersion},
-		}
-	}
-	resp, err := vaultclient.Do(ctx, client, req)
-	if err != nil {
-		return "", fmt.Errorf("couldn't read secret %q: %w", secret.ObjectName, err)
+	req := client.NewRequest(method, secretPath)
+	if queryParams != nil {
+		req.Params = queryParams
 	}
-	if resp == nil || resp.Data == nil {
-		return "", fmt.Errorf("empty response from %q, warnings: %v", req.URL.Path, resp.Warnings)
+	if secret.SecretArgs != nil {
+		req.SetJSONBody(secret.SecretArgs)
 	}
 
+	return req, nil
+}
+
+func keyFromData(rootData map[string]interface{}, secretKey string) (string, error) {
+	// Automatically parse through to embedded .data.data map if it's present
+	// and the correct type (e.g. for kv v2).
 	var data map[string]interface{}
-	d, ok := resp.Data["data"]
+	d, ok := rootData["data"]
 	if ok {
 		data, ok = d.(map[string]interface{})
 	}
 	if !ok {
-		data = resp.Data
+		data = rootData
 	}
 
-	content, ok = data[secret.ObjectName].(string)
+	content, ok := data[secretKey].(string)
 	if !ok {
-		return "", fmt.Errorf("failed to get secret content %q as string", data[secret.ObjectName])
+		return "", fmt.Errorf("failed to get secret content %q as string", secretKey)
 	}
 
 	return content, nil
 }
 
+func (p *provider) getSecret(ctx context.Context, client *api.Client, secretConfig config.Secret) (string, error) {
+	var secret *api.Secret
+	var cached bool
+	key := cacheKey{secretPath: secretConfig.SecretPath, method: secretConfig.Method}
+	if secret, cached = p.cache[key]; !cached {
+		req, err := generateRequest(client, secretConfig)
+		p.logger.Debug("Requesting secret", "secretConfig", secretConfig, "method", req.Method, "path", req.URL.Path, "params", req.Params)
+
+		secret, err = vaultclient.Do(ctx, client, req)
+		if err != nil {
+			return "", fmt.Errorf("couldn't read secret %q: %w", secretConfig.ObjectName, err)
+		}
+		if secret == nil || secret.Data == nil {
+			return "", fmt.Errorf("empty response from %q, warnings: %v", req.URL.Path, secret.Warnings)
+		}
+
+		p.cache[key] = secret
+	} else {
+		p.logger.Debug("Secret fetched from cache", "secretConfig", secretConfig)
```

Comment:
I don't think so but this secretConfig doesn't contain any secrets, right? Just method, secret path, params, etc.

Reply:
Yeah that's correct, it's just config that appears under `objects` in the secret provider class.